src/components/utils/TextAndCodeBlock.tsx

```diff
@@ -0,0 +1,110 @@
+import { MouseEvent, useState, useEffect, memo } from "react";
+
+import { Prism as SyntaxHighlighter } from "react-syntax-highlighter";
+import { coy } from "react-syntax-highlighter/dist/esm/styles/prism";
+
+import { Button, Box, Text } from "@chakra-ui/react";
+
+import { CopyIcon } from "@chakra-ui/icons";
+
+import {
+  CODE_BLOCK_DETECT_REGEX,
+  CODE_BLOCK_LANGUAGE_DETECT_REGEX,
+} from "../../utils/constants";
+import { Row } from "../../utils/chakra";
+import { copySnippetToClipboard } from "../../utils/clipboard";
+
+const getNextCodeBlockMatch = (text: string) => {
+  return CODE_BLOCK_DETECT_REGEX.exec(text);
+};
+
+const CopyCodeButton = ({ code }: { code: string }) => {
+  const [copied, setCopied] = useState(false);
+
+  const handleCopyButtonClick = async (e: MouseEvent) => {
+    e.stopPropagation(); // Prevent this from triggering edit mode in the parent.
+
+    const result = await copySnippetToClipboard(code);
+    if (result) {
+      setCopied(true);
+    }
+  };
+
+  useEffect(() => {
+    if (copied) {
+      const timer = setTimeout(() => setCopied(false), 2000);
+      return () => clearTimeout(timer);
+    }
+  }, [copied]);
+
+  return (
+    <Button
+      onClick={handleCopyButtonClick}
+      size="sm"
+      variant="ghost"
+      _hover={{ background: "#EEEEEE" }}
+    >
+      <CopyIcon boxSize={4} mr={1} /> {copied ? "Copied!" : "Copy Code"}
+    </Button>
+  );
+};
+
+const TitleBar = ({ language, code }: { language?: string; code: string }) => {
+  return (
+    <Row
+      mainAxisAlignment="flex-start"
+      crossAxisAlignment="stretch"
+      expand
+      justifyContent="space-between"
+      alignItems="center"
+      padding="5px 10px"
+      backgroundColor="#f5f5f5"
+      borderBottom="1px solid #eee"
+      borderRadius="6px"
+    >
+      {language ? <Box>{language}</Box> : <Box>plaintext</Box>}
+      <CopyCodeButton code={code} />
+    </Row>
+  );
+};
+
+export const TextAndCodeBlock = memo(({ text }: { text: string }) => {
+  const match = getNextCodeBlockMatch(text);
```

Comment:
Can this operation be hard to perform? Since we could have a large prompt, (32k tokens with gpt4 right?), and this being matched on every render, we could memoize it.

Reply:
the whole component is memoized — so another memo wouldnt do much i believe